### PR TITLE
Condition reorder

### PR DESCRIPTION
**Pull request by JoelMcKinnon**
Drag and drop re-order for conditions within a condition set. Clicking on grab handle within condition title allows dragging up or down in the list to re-order. If the drop hint target is valid (not same index as condition being moved) the drop hint becomes visible and expanded in height. Dropping on a visible target resets the condition list order.

Still not implemented: fix for scroll bar isolated to condition list.

### CODE REVIEW

File: src/plugins/condition/components/ConditionCollection.vue [hunk 1]
```diff
@@ -28,25 +28,34 @@
...
+                             @dragleave="dragLeave"
+                             @dragover.prevent
+                        ></div>
+                        <ConditionEdit :condition-identifier="conditionIdentifier"
+                                       :current-condition-identifier="currentConditionIdentifier"
+                                       :condition-index="index"
+                                       @update-current-condition="updateCurrentCondition"
+                                       @remove-condition="removeCondition"
+                                       @condition-result-updated="handleConditionResult"
+                                       @set-move-index="setMoveIndex"
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionCollection.vue]: You don't need this `set-move-index` event.
In ConditionEdit you can set the startIndex by doing this:
`e.dataTransfer.setData("moveIndex", this.conditionIndex.toString());`

And then retrieve it in ConditionCollection.dropCondition by doing this:
`let moveIndex = e.dataTransfer.getData('startIndex');`
Then you can cast it to a Number if needed.

File: src/plugins/condition/components/ConditionEdit.vue [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
      :class="['widget-condition', { 'widget-condition--current': currentConditionIdentifier && (currentConditionIdentifier.key === conditionIdentifier.key) }]"
 >
     <div class="title-bar">
-        <span
-            class="c-c__menu-hamburger"
-            :class="{ 'is-enabled': !condition.isDefault }"
-            @click="expanded = !condition.expanded"
+        <span class="c-c__menu-hamburger"
+              :class="{ 'is-enabled': !condition.isDefault }"
+              :data-condition-index="conditionIndex"
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionEdit.vue]: Is the `:data-condition-index` property needed? Could you use `this.conditionIndex` that has been passed into props directly in the dragstart method?

File: src/plugins/condition/components/ConditionEdit.vue [hunk 3]
```diff
@@ -205,6 +211,11 @@ export default {
         this.persist();
     },
     methods: {
+        dragStart(e) {
+            e.dataTransfer.effectAllowed = "copyMove";
+            e.dataTransfer.setDragImage(e.target.closest('.c-c-container__container'), 0, 0);
+            this.$emit('set-move-index', Number(e.target.getAttribute('data-condition-index')));
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionEdit.vue]: See comment above about not needing the `set-move-index` event.

File: src/plugins/condition/components/DropHint.vue [hunk 4]
```diff
@@ -0,0 +1,83 @@
+/*****************************************************************************
+ * Open MCT, Copyright (c) 2014-2018, United States Government
```
Comment (shefalijoshi) [inline src/plugins/condition/components/DropHint.vue]: Change year to `2020`

File: src/plugins/condition/components/ConditionCollection.vue [hunk 5]
```diff
@@ -93,6 +104,50 @@ export default {
         }
     },
     methods: {
+        setMoveIndex(index) {
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionCollection.vue]: You don't need this method. See comment above about not using the `set-move-index` event.

File: src/plugins/condition/components/ConditionCollection.vue [hunk 6]
```diff
@@ -72,7 +82,8 @@ export default {
             parentKeyString: this.openmct.objects.makeKeyString(this.domainObject.identifier),
             conditionCollection: [],
             conditions: [],
-            currentConditionIdentifier: this.currentConditionIdentifier || {}
+            currentConditionIdentifier: this.currentConditionIdentifier || {},
+            moveIndex: null
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionCollection.vue]: You don't need this `moveIndex` property. See comment above about not using the `set-move-index` event.

File: src/plugins/condition/components/ConditionCollection.vue [hunk 7]
```diff
@@ -93,6 +104,50 @@ export default {
         }
     },
     methods: {
+        setMoveIndex(index) {
+            this.moveIndex = index;
+        },
+        dropCondition(e) {
+            let targetIndex = Array.from(document.querySelectorAll('.c-c__drag-ghost')).indexOf(e.target);
+            if (targetIndex > this.moveIndex) { targetIndex-- }
```
Comment (shefalijoshi) [inline src/plugins/condition/components/ConditionCollection.vue]: See comment above about using `event.dataTransfer.getData(...)` to get the moveIndex instead of using `this.moveIndex` here.

File: src/plugins/condition/components/condition.scss [hunk 8]
```diff
@@ -1,8 +1,6 @@
-
-
 .widget-condition {
     background-color: #eee;
-    margin: 0 0 0.6em;
+    margin: 0 0 5px;
```
Comment (shefalijoshi) [inline src/plugins/condition/components/condition.scss]: Prefer the use of `em` instead of `px`. Was this change necessary?

File: src/plugins/condition/components/condition.scss [hunk 9]
```diff
@@ -152,3 +161,15 @@
         content: $glyph-icon-trash;
     }
 }
+
+.c-c__drag-ghost {
+    width: 100%;
+    min-height: 5px;
```
Comment (shefalijoshi) [inline src/plugins/condition/components/condition.scss]: Can we use `em` instead of `px`?

File: src/plugins/condition/components/ConditionCollection.vue (hunk 1 above)
Comment (shefalijoshi) [reply src/plugins/condition/components/ConditionCollection.vue]: N/A - leave as is for now.

File: src/plugins/condition/components/ConditionEdit.vue (hunk 3 above)
Comment (shefalijoshi) [reply src/plugins/condition/components/ConditionEdit.vue]: N/A - leave as is for now.

File: src/plugins/condition/components/ConditionCollection.vue (hunk 5 above)
Comment (shefalijoshi) [reply src/plugins/condition/components/ConditionCollection.vue]: N/A - leave as is for now.

File: src/plugins/condition/components/ConditionCollection.vue (hunk 7 above)
Comment (shefalijoshi) [reply src/plugins/condition/components/ConditionCollection.vue]: N/A - leave as is for now.

File: src/plugins/condition/components/ConditionCollection.vue (hunk 6 above)
Comment (shefalijoshi) [reply src/plugins/condition/components/ConditionCollection.vue]: N/A - leave as is for now.

File: src/plugins/condition/components/ConditionEdit.vue (hunk 2 above)
Comment (JoelMcKinnon) [reply src/plugins/condition/components/ConditionEdit.vue]: done

File: src/plugins/condition/components/DropHint.vue (hunk 4 above)
Comment (JoelMcKinnon) [reply src/plugins/condition/components/DropHint.vue]: DropHint.vue isn't used and shouldn't have been part of this PR. I've removed it. I went ahead and added the current legal comment header for all vue components in condition directory

File: src/plugins/condition/components/condition.scss (hunk 9 above)
Comment (JoelMcKinnon) [reply src/plugins/condition/components/condition.scss]: done

File: src/plugins/condition/components/condition.scss (hunk 8 above)
Comment (JoelMcKinnon) [reply src/plugins/condition/components/condition.scss]: done

File: src/plugins/condition/components/ConditionCollection.vue (hunk 1 above)
Comment (psarram) [reply src/plugins/condition/components/ConditionCollection.vue]: @JoelMcKinnon Please use camelCase for event names for consistency. There are only couple of places in the source code (flexible layout and date picker) where it's not done that way, which we'll need to update at some point.

Comment (JoelMcKinnon) [reply src/plugins/condition/components/ConditionCollection.vue]: VueJS docs say you should always use kebab-case: https://vuejs.org/v2/guide/components-custom-events.html#Event-Names

I tried changing to camelCase and the events didn't work.

Comment (psarram) [reply src/plugins/condition/components/ConditionCollection.vue]: They are recommending it,  but it doesn't mean camelCase event names don't work.

We need to discuss this with the team to decide which way to go.
